### PR TITLE
perf: Improve Detected labels API

### PR DESCRIPTION
**Pull request by shantanualsi**
**What this PR does / why we need it**:

- **Remove usage of regex to improve performance**
- **Enable query splitting for detected labels**
- **Change detected labels implementation to use hyperloglog**

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/querier/querier.go [hunk 1]
```diff
@@ -1031,28 +1048,15 @@ func (q *SingleTenantQuerier) Patterns(ctx context.Context, req *logproto.QueryP
...
-	pattern := `^(?:(?:[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})|(?:(?:\{)?[0-9a-fA-F]{8}(?:-?[0-9a-fA-F]{4}){3}-?[0-9a-fA-F]{12}(?:\})?)|(\d+(?:\.\d+)?))$`
-
-	re := regexp.MustCompile(pattern)
-
 	for _, v := range values {
-		if !re.MatchString(v) {
-			return false
+		_, err := strconv.ParseFloat(v, 64)
+		if err != nil {
+			_, err = uuid.Parse(v)
```
Comment (trevorwhitney) [inline pkg/querier/querier.go:1056]: much cleaner than the regex pattern

File: pkg/querier/querier.go [hunk 2]
```diff
@@ -979,40 +978,58 @@ func (q *SingleTenantQuerier) DetectedLabels(ctx context.Context, req *logproto.
...
+	var detectedLabels []*logproto.DetectedLabel
+	for k, v := range dlMap {
+		sketch, err := v.sketch.MarshalBinary()
+		if err != nil {
+			// TODO: add log here
+			continue
+		}
+		detectedLabels = append(detectedLabels, &logproto.DetectedLabel{
+			Label:       k,
+			Cardinality: v.Estimate(),
```
Comment (trevorwhitney) [inline pkg/querier/querier.go:1028]: yeah, good call here, we hopefully don't care about this but if there's only 1 split we'll need the estimate here

File: pkg/ingester/ingester_test.go [hunk 3]
```diff
@@ -833,8 +833,8 @@ func TestIngester_GetDetectedLabels(t *testing.T) {
 	require.NoError(t, err)
 
 	res, err := i.GetDetectedLabels(ctx, &logproto.DetectedLabelsRequest{
-		Start: &[]time.Time{time.Now().Add(11 * time.Nanosecond)}[0],
-		End:   nil,
+		Start: []time.Time{time.Now().Add(11 * time.Nanosecond)}[0],
+		End:   []time.Time{time.Now().Add(11 * time.Nanosecond)}[0],
```
Comment (trevorwhitney) [inline pkg/ingester/ingester_test.go:837]: Why the same time?

File: pkg/querier/queryrange/roundtrip.go [hunk 4]
```diff
@@ -307,11 +308,38 @@ func NewDetectedLabelsTripperware(cfg Config, opts logql.EngineOpts, logger log.
...
+
+			resp, ok := res.(*DetectedLabelsResponse)
+			if !ok {
+				return res, nil
+			}
+
+			var result []*logproto.DetectedLabel
+
+			for _, dl := range resp.Response.DetectedLabels {
+				if dl.Cardinality > 2 && dl.Cardinality < 50 {
```
Comment (trevorwhitney) [inline pkg/querier/queryrange/roundtrip.go:332]: I still get confused by our middlewares, so just to make sure, this is happening after we `MergeLabels` right? I wouldn't want to omit results based on partial cardinality of a subquery.

File: pkg/storage/detected/labels.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		l, ok := mergedLabels[label.Label]
+		if !ok {
+			unmarshaledLabel, err := unmarshalDetectedLabel(label)
+			if err != nil {
+				return nil, err
+			}
+			mergedLabels[label.Label] = unmarshaledLabel
+		}
+
+		if ok {
```
Comment (trevorwhitney) [inline pkg/storage/detected/labels.go:47]: could be an `else`, might read a bit better?

**Review by trevorwhitney — APPROVED** [review]
looks good, just a few small things.

File: pkg/querier/querier.go (hunk 2 above)
Comment (shantanualsi) [reply pkg/querier/querier.go:1028]: sorry didn't get the issue.. even if there's a 1 split, the estimate will be populated here right?

File: pkg/ingester/ingester_test.go (hunk 3 above)
Comment (shantanualsi) [reply pkg/ingester/ingester_test.go:837]: will change this, missed updating it after changing to value semantics

File: pkg/storage/detected/labels.go (hunk 5 above)
Comment (shantanualsi) [reply pkg/storage/detected/labels.go:47]: lol yes!   😆

File: pkg/querier/queryrange/roundtrip.go (hunk 4 above)
Comment (shantanualsi) [reply pkg/querier/queryrange/roundtrip.go:332]: yes confirmed again this is executed after MergeLabels. The line `res, err := rt.Do(ctx, req)` is before everything else

**Pull request merged**